consensus/src/simplex/mod.rs

```diff
@@ -4939,4 +4941,742 @@ mod tests {
+        fn recipients<P: Clone>(self, view: View, participants: &[P]) -> TwinRecipients<P> {
+            let n = participants.len();
+            match self {
+                TwinPartition::View => {
+                    let split = view.get() as usize % n;
+                    let (primary, secondary) = participants.split_at(split);
+                    TwinRecipients {
+                        twin_primary: primary.to_vec(),
+                        twin_secondary: secondary.to_vec(),
+                    }
+                }
+                TwinPartition::Fixed(split) => {
+                    let (primary, secondary) = participants.split_at(split);
+                    TwinRecipients {
+                        twin_primary: primary.to_vec(),
+                        twin_secondary: secondary.to_vec(),
+                    }
+                }
+                TwinPartition::Broadcast => TwinRecipients {
+                    twin_primary: participants.to_vec(),
+                    twin_secondary: participants.to_vec(),
+                },
+                TwinPartition::Isolate(idx) => TwinRecipients {
+                    twin_primary: vec![participants[idx].clone()],
+                    twin_secondary: participants
+                        .iter()
+                        .enumerate()
+                        .filter(|(i, _)| *i != idx)
+                        .map(|(_, p)| p.clone())
+                        .collect(),
+                },
+                TwinPartition::Shuffle => {
+                    let mut rng = StdRng::seed_from_u64(view.get());
+                    let mut shuffled: Vec<_> = participants.to_vec();
+                    shuffled.shuffle(&mut rng);
+                    let split = rng.gen_range(0..n);
+                    let (primary, secondary) = shuffled.split_at(split);
+                    TwinRecipients {
+                        twin_primary: primary.to_vec(),
+                        twin_secondary: secondary.to_vec(),
+                    }
+                }
+            }
+        }
+
+        /// Determines which twin should receive a message from a given sender at a given view.
+        fn route<P: Clone + PartialEq>(
+            self,
+            view: View,
+            sender: &P,
+            participants: &[P],
+        ) -> SplitTarget {
+            let recipients = self.recipients(view, participants);
+            let in_primary = recipients.twin_primary.contains(sender);
+            let in_secondary = recipients.twin_secondary.contains(sender);
+            match (in_primary, in_secondary) {
+                (true, true) => SplitTarget::Both,
+                (true, false) => SplitTarget::Primary,
+                (false, true) => SplitTarget::Secondary,
+                (false, false) => panic!("sender not in any partition"),
+            }
+        }
+    }
+
+    #[test]
+    fn test_twin_partition_view_split_varies_by_view() {
```

Comment:
Figured it was worthwhile to add tests for the recipients/route logic because if that's borked, we won't be testing what we think we're testing...